### PR TITLE
Portable random number generation

### PR DESCRIPTION
**Pull request by lidavidm**
Pulls the Clang libcxx implementation of `uniform_int_distribution` and `uniform_real_distribution` (MIT license) into the engine itself, hopefully giving us deterministic random number generation across platforms.

### CODE REVIEW

File: environment/core/util/distributions.hpp [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        // Search from LSB to MSB for first set bit.
+        // Returns zero if no set bit is found.
+        if (_BitScanReverse(&where, mask))
+            return 31 - where;
+        return 32; // Undefined Behavior.
+#endif
+    }
+
+    static unsigned long __clz(unsigned long __x) {
+#ifndef _LIBCPP_COMPILER_MSVC
```
Comment (Snaar) [inline environment/core/util/distributions.hpp]: Can you add this define to the make.bat, make2017.bat, make2017_32.bat? I promise I'll turn them into single bat file soon, been working on cleaning that up :P

File: environment/core/util/distributions.hpp [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        return static_cast<result_type>(__u + __p.a());
+    }
+
+    template<class _RealType, std::size_t __bits, class _URNG>
+    _RealType
+    generate_canonical(_URNG& __g)
+    {
+        const std::size_t _Dt = std::numeric_limits<_RealType>::digits;
+        const std::size_t __b = _Dt < __bits ? _Dt : __bits;
+#ifdef _LIBCPP_CXX03_LANG
```
Comment (Snaar) [inline environment/core/util/distributions.hpp]: This looks safe for me to leave as is undefined anywhere. Just fyi for completeness.

**Comment by Janzert** [comment]
This currently misses the distribution at line 196 of main.cpp where map size is set. It also doesn't update the asteroid cluster generator, but that is unused at this point so probably doesn't matter.

File: environment/core/util/distributions.hpp [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+    }
+
+    static unsigned long long __clz(unsigned long long __x) {
+#ifndef _LIBCPP_COMPILER_MSVC
+        return static_cast<unsigned long long>(__builtin_clzll(__x));
+#else
+        unsigned long where;
+        // BitScanReverse scans from MSB to LSB for first set bit.
+        // Returns 0 if no set bit is found.
+#if defined(_LIBCPP_HAS_BITSCAN64)
```
Comment (Snaar) [inline environment/core/util/distributions.hpp]: Ok to leave undefined. We do support building for 32 bit windows, which lacks _BitScanReverse64. Don't care about performance implications here.

File: environment/core/util/distributions.hpp (hunk 1 above)
Comment (lidavidm) [reply environment/core/util/distributions.hpp]: Done, thanks for pointing that out.

File: environment/core/util/distributions.hpp (hunk 2 above)
Comment (lidavidm) [reply environment/core/util/distributions.hpp]: I went ahead and took out that branch, since we don't need it anyways.

**Comment by lidavidm** [comment]
@Janzert thanks for pointing that out, updated that as well.

**Comment by Snaar** [comment]
Ok, had to do 7834a13c5e1e9dcbdfb475a11487d34ca4fe9695 and 8eb11f6c5f9072ba3117bfb3e381416ccbfdef70 to make it compile and work on windows. It produces identical results now for same seed as linux version.

**Comment by Snaar** [comment]
It generates quite some warnings on windows now - I'll clean it up later.

**Pull request merged**